### ISSUE TITLE
Documentation does not mention Windows binary of Skopeo.

### ISSUE THREAD

**Issue by tandeday** [containers/skopeo#715]
A quick read of the documentation did not immediately reveal how I got a Windows (10 if that matters) binary of skopeo for me to use against our OpenShift cluster.  I do not have Go experience so I didn't just go ahead and build it.

It would be nice if the documentation clearly stated:

* If a binary is available for download (probably isn't).  
* How to create one if possible? (along with a difficulty estimage).
* If not possible within reason, then explicitly mention that.

Thanks

/Thorbjørn

**Comment by vrothberg** [comment]
Hi @m86194, thanks for opening this issue.

We don't offer binaries for any OS. Skopeo is (build) tested against Linux and Mac OS but I am not sure how far we can come on Windows. I don't have access to a Windows machine for testing.

@mtrmac @rhatdan WDYT?

**Comment by mtrmac** [comment]
We do test Windows cross-compilation of c/image, so it is probably buildable, but I wouldn’t be at all surprised if Skopeo were a pretty bad Windows citizen (WRT configuration file paths and the like), and not extremely surprised if it were broken due to different path syntax or the like (though there was at least one Windows-based user of the c/image library at one point).

`make binary-local` would be the default way to try to build it, adding some `BUILDTAGS` might be necessary.

Frankly, making Windows work, and work smoothly, and having everything well documented, would be mostly dependent on someone interested contributing that. We are definitely interested, at least enough not to sabotage that :) , but lack both access and time to do it ourselves.

Same for providing pre-built binaries; on Linux it is more or less an intentional choice not to, because a single binary is not portable to various Linuxes; on Windows it might be more practical, but again dependent on someone integrating that into the current build system.

**Comment by xphyr** [comment]
SO, it very much seems possible. I was able to build a Windows binary just this evening (completely unrelated to @m86194 request). I made a small tweak to the Makefile (created a Windows build target) and it does compile. I will test it out tomorrow at work on a Windows machine and see if its usable. If you are curious about the addition to the Makefile take a look at my fork of this project.

**Comment by m86194** [comment]
First of all, please notice I am not requesting a windows build (which would be nice, though) but that the _documentation_ is clearer about what can be expected from a new user coming to the project.  (I need a dockerless image mirrorer) because it is slowly showing up in google searches for this kind of utility.

I fully understand that you have limited resources and this is to be considered an internal Redhat tool which just happens to be useful to others, instead of a fully funded product.    I am just letting you know that people like me want to know this pretty quickly when triaging posibillities.

@xphyr this sounds very interesting.  If it is usable, I'd like to have a look at it.  Thanks.

**Comment by rhatdan** [comment]
Well, first of all this is not an Internal Red Hat Tool.  Skopeo is a fully open source project and we welcome it running on any platform.  Linux, MAC, Windows ...

If we get contributors working on the Windoews and MAC Ports that would be great.  The primary developers right now are Linux users and therefore it gets the most attention.

We are working to make Podman easier to consume on Windows and MAC, so I don't see a reason why we would not want to get Skopeo easily available on those platforms as well.

**Comment by xphyr** [comment]
Over the weekend I tested the functionality of the Windows and OSX builds, using the Make file in my fork and I documented the results in the readme as well. Generally speaking, inspect, copy & delete work, as does the various storage subsystems except those specifically tied to the linux operating system (such as "ostree support"). It looks like this was known for OSX, but not for Windows.

If there is interest, I would be happy to work on a binary release process for OSX/Windows so that others could leverage this tool on alternative platforms without the need for installing Go and Make. My first thought would be to either use either:

* goreleaser as part of the Travis CI build process
* the native TravisCI release module.

Please let me know if this is something you would like to accept into the project and I will work on putting it together.

**Comment by rhatdan** [comment]
Yes, I think we are very interested in this.

**Comment by vrothberg** [comment]
Absolutely, thanks for looking into this, @xphyr !

**Comment by xphyr** [comment]
I have been able to make relatively few changes to the build process and I now have a travis build that runs against Linux, OSX and Windows and passes all tests successfully.  I am now working on how to get travis to create a github "release" so that we can upload a compiled binary for windows and osx. The on-going work is in my fork if anyone is interested in seeing progress or has any feedback.

**Comment by mtrmac** [comment]
Great!

Some quick observations looking at that branch (without checking to verify that what I’m talking about is practical right now, it may just be my ignorance):
- If possible, I’d like to minimize proliferation of build targets, and rule duplication; e.g. the primary `make binary-local` target, when run on macOS/Windows, should automatically produce a working macOS/Windows binary, and we shouldn’t need to tell the users to run different commands on different systems. (Your work already moves in this direction, with the `OS`/`name` conditionals). cross-OS build targets may still be useful, but they are distinctly secondary to native builds, and ideally should reuse, not duplicate, the primary build rules (e.g. a `binary-local-windows` target could just be `make binary-local OS=Windows_NT` or something like that).
- We’re, by default, emphasizing the non-container builds (`make binary-local`); if they work, and they are the only ones used by CI, I’m not sure we even need the container ones to be added; they bring little value and are somewhat likely to be broken without anyone noticing.
- At least on Linux, the default is to use dynamic, not static, linking, unless there is a good reason (dynamic linking is preferred to allow independent updates of the called libraries, including security fixes). I’m not sure whether this applies to macOS/Windows as well (I don’t know all that much about them) – it may well not be the same – but I’d like this to be, eventually, at least explicitly discussed.

**Comment by xphyr** [comment]
Thanks @mtrmac for the feedback.  I agree the multiple build targets was getting a bit much, and I really like your solution.  I will look to incorporate this into the work I am doing.
I will need to look more into static vs dynamic linking on macOS/Windows. I will admit ignorance here and say that I have only ever built static binaries for these OSes.  One of the things that has made building the Windows binary more challenging is the CGO requirement for gpgme in the containers/image library. I may look into this next, just to see if there is a solution to this problem as my next project to tackle ...

**Comment by xphyr** [comment]
End of the weekend update, my football team lost, but I was able to clean up the multiple build targets, and make "make" do something much more sane. I am now detecting the need for "containers_image_openpgp" dynamically, similar to the detection of OSTREE, LIBDM and BTRFS tags. This has us back to just "binary-local" and "binary-local-static" targets. I have also backed out the explicit requirement to build static and the BUILD_OS is auto-detected on start, but can be over-ridden for testing purposes.

I also had success in publishing a Windows and OSX build binary to my fork. I have done a preliminary test of the Windows binary and it works. If you check my fork under releases you will see a Windows and OSX (darwin) release file for version 0.1.40-xphyr.

I am wondering if I should tackle #580 while I am in the Makefile making changes.  After reading 580, as well as the comments above about not needing containers should I attempt to strip the container build targets from the Makefile, or do we want to do that as a separate PR (if at all).

**Comment by sebastianwebber** [comment]
hello all!

do you guys still have an interest to use GoReleaser to make the releases? Can I open a PR?

**Comment by TomSweeneyRedHat** [comment]
@sebastianwebber we're interested, @xphyr have you made any further progress?

**Comment by sebastianwebber** [comment]
hey @TomSweeneyRedHat and @xphyr I added a draft PR but i need some info to finish.

**Comment by xphyr** [comment]
Hello all, sorry I must be honest, I completely forgot about this.  We got a version working for windows using my fork and then I had to move on... my fork is very outdated so I should really go back and take a look at if my make file changes would still work.  Using GoReleaser is a good idea and I could take a look at what what @sebastianwebber has put together for it ... the question becomes does the skopeo project want to take on a different build process or stick with "make"...  (BTW I am now a Red Hatter (yeah!) so I can hopefully see this through to completion now. )

**Comment by TomSweeneyRedHat** [comment]
@xphyr welcome to Red Hat!  @mtrmac and @vrothberg 's call, but I'd like to stick with make if we can.  But if it turns out it's not possible to pull off a windows build that way, I'd not be against a separate build.

@xphyr did you want to tackle #580?  If not, we probably can twist @parkervanroy 's arm and ask him to take a look.

**Comment by TomSweeneyRedHat** [comment]
and I'll try using Parker's real GitHub handle this time, @ParkerVR.  Parker please see the above comment.

**Comment by ParkerVR** [comment]
@TomSweeneyRedHat happy to look into this, but I don’t have much bandwidth this week and am moving next week- can get into it after that.

**Comment by xphyr** [comment]
@TomSweeneyRedHat thanks for the welcome .. who knew a year ago I would be here!  I will take a look at both the windows build as well as tackling #580 I have not looked at this in a while, so will take me a few minutes to get back up to speed. I will status this issue later in the week with where I have gotten.

**Comment by rhatdan** [comment]
@ParkerVR Stick to the shortnames work.

**Comment by rhatdan** [comment]
@ashley-cui PTAL

**Comment by awakecoding** [comment]
Is this still being worked on? I'm not a Go dev, but I just tried running "go build" in "skopeo\cmd\skopeo", and it fails with the following:

```
# github.com/mtrmac/gpgme
cgo: exec gcc: exec: "gcc": executable file not found in %PATH%
# github.com/miekg/pkcs11
cgo: exec gcc: exec: "gcc": executable file not found in %PATH%
# github.com/containers/storage/pkg/unshare
cgo: exec gcc: exec: "gcc": executable file not found in %PATH%
```

I'm surprised that Go packages try calling gcc directly, but has anyone managed to get a Windows build, and if so, how?

**Comment by lsm5** [comment]
@TomSweeneyRedHat @rhatdan does a pre-packaged windows binary exist?

**Comment by awakecoding** [comment]
I just tried building podman as well and it fails for exactly the same packages, reading a bit more about it, it looks like they use cgo to build some C code, and cgo apparently always calls gcc, even on Windows, with the only solution being to install MinGW and make sure gcc.exe is available in the PATH. I guess once someone has gcc.exe available in Windows it becomes possible to go much further in the build process. Is my analysis correct, or can someone use clang-llvm, or even MSVC as the compiler on Windows for cgo?

**Comment by lsm5** [comment]
@awakecoding couple of questions:

1. Is this something that can't be handled by a linux binary + WSL2 (just thinking out loud, no idea myself)

2. Would you like to own this issue?

**Comment by xphyr** [comment]
I just took a look at this, and it looks like you can build a windows binary leveraging the make file.  

I am using a OSX box to test, but if you run `make local-cross` you will end up with a compiled binary for OSX, linux.arm and windows.  

You can also run `make bin/skopeo.windows.amd64.exe` and get just a windows binary.

There are caveats to the Windows binary (things like btrfs support are not available for obvious reasons).

**Comment by awakecoding** [comment]
@lsm5 @xphyr this isn't the first Go project where it appears difficult to get the thing built *on* Windows. If it's possible to cross-compile it, maybe add a prebuilt binary so people can start experimenting with it? It would still be quite useful to be able to build on Windows, without cross-compiling.

I don't have much time to spend on this, I was curious to try it out on Windows, and there was no binary around so I thought I may try building it. I got a little bit further, but I'm running into other issues now. I'll leave my notes if someone wants to give it a shot:

Install mingw-w64 and gpg4win:

```
choco install mingw
choco install gpg4win
```

Copy and rename the prebuilt gpg4win files alongside skopeo:

```
cd skopeo\cmd\skopeo
$GPG_DIR = "${Env:ProgramFiles(x86)}\GnuPG"
Copy-Item "$GPG_DIR\include" . -Recurse -Force
Copy-Item "$GPG_DIR\lib" . -Recurse -Force
Copy-Item "$GPG_DIR\bin\*.dll" ".\lib"
Get-ChildItem ".\lib\*.imp" | Rename-Item -NewName { $_.Name -Replace '.imp','.lib' }
```

gpg4win is unfortunately prebuilt for 32-bit Windows, and building it from source looks very painful. I would rather use the 64-bit version if someone has it. Anyway, you can set the environment variables to attempt building it:

```
$Env:GOARCH="386"
$env:GOOS="windows"
$Env:CGO_CFLAGS="-m32 -I$PWD\include"
$Env:CGO_LDFLAGS="-L$PWD\lib" 
```

The include files and the libraries will be found for gpgme, but then it fails later on:

```
# github.com/containers/image/v5/signature
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:16:16: undefined: gpgme.Context
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:66:44: undefined: gpgme.Context
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:67:14: undefined: gpgme.New
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:71:27: undefined: gpgme.ProtocolOpenPGP
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:75:28: undefined: gpgme.ProtocolOpenPGP
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:97:20: undefined: gpgme.NewDataBytes
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:126:20: undefined: gpgme.NewDataBytes
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:131:18: undefined: gpgme.NewDataWriter
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:135:25: undefined: gpgme.Key
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:135:61: undefined: gpgme.SigModeNormal
C:\go\pkg\mod\github.com\containers\image\v5@v5.10.2-0.20210219165336-d21b4da7f368\signature\mechanism_gpgme.go:135:61: too many errors
```

I used some tricks from this old issue: https://github.com/proglottis/gpgme/issues/14

Maybe someone can go further than I did. If we could just have a proper build procedure for all the cgo packages on Windows, it could unblock native compilation for a lot of projects. From what I've seen, gpgme and a couple of packages used in skopeo are also used in many other projects like podman, etc.

**Comment by varyumin** [comment]
Hello everyone! Has anyone been built skopeo for Windows?

**Comment by lsm5** [comment]
> Hello everyone! Has anyone been built skopeo for Windows?

Not to my knowledge. Contributions welcome!

**Comment by lsm5** [comment]
I wonder if the podman windows packaging could be reused for skopeo in case anyone is interested to take that up.

**Comment by Rickaro** [comment]
There are two (may be more) options for build skopeo for win - with gpgme (native build) or without gpgme (native or cross-compilation)
<details>
<summary>Build without gpgme </summary>
Resulting binary could copy, inspect, and verify signatures, but not create new one.

It could be easily  created with cross compilation
```
root@my-shell-77f597f67f-q28x4:/# cat /etc/issue
Ubuntu 20.04.3 LTS \n \l

# apt update && apt install -y mingw-w64 curl golang make
# curl -L https://github.com/containers/skopeo/archive/refs/tags/v1.4.1.tar.gz -o - | tar xzv
# cd skopeo-1.4.1 && make bin/skopeo.windows.386.exe && make bin/skopeo.windows.amd64.exe
```

This binaries completely suits my needs, but report Error when I try to create signature
```
skopeo standalone-sign manifest.json docker.io/some ABB77339FD44B54ADA502424B4EEDCEE062C5D3B -o out.sig
time="2021-09-11T15:02:57+03:00" level=fatal msg="Error creating signature: signing is not supported in github.com/containers/image built with the containers_image_openpgp build tag" 
```
Same binaries could be created natively with mingw (see build process details bellow,  `make` should be called after unpacking sources)
```
tar xzvf skopeo.tgz && cd skopeo-1.4.1

mingw32-make bin/skopeo.windows.386.exe
# or
mingw32-make bin/skopeo.windows.amd64.exe
```
</details>
<details>
<summary>Build with gpgme</summary>

I've use VM with Win10 **x64**

```
# Install packages
choco install -y gpg4win msys2 golang
```

Let's go to bash:
- For x86_64 compilation start `C:\tools\msys64\mingw64.exe`
- For i686 compilation start `C:\tools\msys64\mingw32.exe`

```
# Here we have environment vars
# MSYSTEM="MINGW32"
# MSYSTEM_CARCH="i686"
# MSYSTEM_CHOST="i686-w64-mingw32"

# MSYSTEM="MINGW64"
# MSYSTEM_CARCH="x86_64"
# MSYSTEM_CHOST="x86_64-w64-mingw32"

# Install build tools 
pacman -S --noconfirm curl mingw-w64-${MSYSTEM_CARCH}-gcc mingw-w64-${MSYSTEM_CARCH}-make

# Download and unpack src
curl -L https://github.com/containers/skopeo/archive/refs/tags/v1.4.1.tar.gz -o skopeo.tgz
tar xzvf skopeo.tgz && cd skopeo-1.4.1

# Here we need edit Makefile and comment line 11, starts with GPGME_ENV, in order to be able to set CGO variables
sed -i -e '11s/^GPGME_ENV/#GPGME_ENV/' Makefile

# additional patch should be applied on go_gpgme.h
# https://github.com/proglottis/gpgme/issues/14#issuecomment-298389837
#
# It is required for i686 compilation. It could be applied on x86_64, build will not fail.
if [ "$MSYSTEM" == "MINGW32" ]; then
  sed -i -e '16s/\boff_t/gpgme_off_t/g' ./vendor/github.com/mtrmac/gpgme/go_gpgme.h
fi

# Create symlink to gpg4win directory to omit spaces in path
ln -s /c/Program\ Files\ \(x86\)/Gpg4win /usr/local/gpg4win

# Next - we will use dll files for linking
# i686 located at /usr/local/gpg4win/bin
# x86_64 located at /usr/local/gpg4win/bin_64
# dll names should be changed for ld: libassuan6-0.dll -> libassuan.dll 

# Create symlinks for libraries, so linker can found them
if [ "$MSYSTEM" == "MINGW32" ]; then
  dll_list="libassuan-0.dll libgpg-error-0.dll libgpgme-11.dll"
  dll_path=/usr/local/gpg4win/bin

  export GOARCH=386
fi

if [ "$MSYSTEM" == "MINGW64" ]; then
  dll_list="libassuan6-0.dll libgpg-error6-0.dll libgpgme-11.dll"
  dll_path=/usr/local/gpg4win/bin_64
  
  export GOARCH=amd64
fi

for lib in $dll_list; do
  name=$(echo $lib | sed -r 's/[[:digit:]]?-[^-]+?(\.dll)/\1/')
  ln -s $dll_path/$lib $dll_path/$name
done

# Set path to golang executables
export PATH=$PATH:/c/Program\ Files/Go/bin

CGO_ENABLED=1 \
  CC=${MSYSTEM_CHOST}-gcc CXX=${MSYSTEM_CHOST}-g++ \
  CGO_CFLAGS="-I/usr/local/gpg4win/include" CGO_LDFLAGS="-L$dll_path" \
  mingw32-make.exe bin/skopeo

[ -d /c/tools/skopeo ] && rm -rf /c/tools/skopeo

mkdir /c/tools/skopeo
cp bin/skopeo /c/tools/skopeo/skopeo.exe

# 64bit version is linked dynamically, so we need dlls near binary
# 32bit - linked statically
if [ "$MSYSTEM" == "MINGW64" ]; then
  for lib in $dll_list gpgme-w32spawn.exe; do
    cp $dll_path/$lib /c/tools/skopeo/
  done
fi

# and leave bash
exit
```

msys2 could be removed
```
choco uninstall msys2
rd /S /Q C:\tools\msys64
```

After that we have in result in `C:\tools\skopeo` and could try it
```
rem Create key
gpg --gen-key
rem And sign manifest
C:\tools\skopeo\skopeo.exe standalone-sign manifest.json docker.io/some -o out.sig ABB77339FD44B54ADA502424B4EEDCEE062C5D3B
```
</details>

Some notes:
For native build on 32bit OS - msys2 32bit is not present in chocolatly, and need to be installed manually.
For 32bit version dlls from gnupg could be used, and gpg4win not required
For 32bit version Instead of dll - .imp files from lib folder could be used
I did not use whole skopeo functionality, so can't test all features of resulting builds

Hope it helps

**Comment by mtrmac** [comment]
@lsm5 #1464 refers to this as “an open feature request”, so I’ll reopen; please speak up if that’s incorrect.

**Comment by Rickaro** [comment]
@gbraad

for building 64bit version
<details>
<summary>appveyor.yml</summary>

```
version: 1.0.{build}
image: Visual Studio 2015
install:
- ps: |
    $path = [Environment]::GetEnvironmentVariable("path", "machine")
    $newPath = ($path.Split(';') | Where-Object { $_ -eq 'C:\ProgramData\chocolatey\bin' }) -join ';'
    [Environment]::SetEnvironmentVariable("path", $newPath, "machine")
    choco install gpg4win --no-progress
    [Environment]::SetEnvironmentVariable("path", $path, "machine")
build_script:
# APPVEYOR_BUILD_FOLDER=/c/projects/skopeo
# Here we need edit Makefile and comment line 11, starts with GPGME_ENV, in order to be able to set CGO variables
- sed -i -e '11s/^GPGME_ENV/#GPGME_ENV/' Makefile

- |-
  set GPG_INSTALL_DIR=C:\Program Files (x86)\Gpg4win
  set GPG_INSTALL_DLL=%GPG_INSTALL_DIR%\bin_64
  set DIST_DIR=dist
  set GPG_DLL=gpg-dll

- ln -s "%GPG_INSTALL_DIR%" ./
#Next - we will use dll files for linking
#x86_64 located at /c/projects/gpg4win/bin_64
#dll names should be changed for ld: libassuan6-0.dll -> libassuan.dll 
#Create symlinks for libraries, so linker can found them
     
- ps: |
    New-Item -Path $env:DIST_DIR -ItemType Directory
    New-Item -Path $env:GPG_DLL -ItemType Directory

    Get-ChildItem $env:GPG_INSTALL_DIR\bin_64 | `
      Where-Object { $_.Name -match '\d?-[^-]+?(\.dll)$' } | `
      % {
        Copy-Item -Path $env:GPG_INSTALL_DLL\$_ -Destination $env:DIST_DIR
        
        Copy-Item -Path $env:GPG_INSTALL_DLL\$_ -Destination $env:GPG_DLL -PassThru | `
          Rename-Item -NewName { $_.Name -replace '\d?-[^-]+?(\.dll)$','.dll' }
      }

    Copy-Item $env:GPG_INSTALL_DLL\gpgme-w32spawn.exe $env:DIST_DIR

- |-
    set PATH=C:\mingw-w64\x86_64-7.3.0-posix-seh-rt_v5-rev0\mingw64\bin;%PATH%
    set CGO_ENABLED=1
    set GOARCH=amd64
    set MSYSTEM_CHOST=x86_64-w64-mingw32
    set PROJ_DIR=/c/projects/%APPVEYOR_PROJECT_SLUG%
    set CC=%MSYSTEM_CHOST%-gcc 
    set CXX=%MSYSTEM_CHOST%-g++ 
    set CGO_CFLAGS=-I%PROJ_DIR%/Gpg4win/include
    set CGO_LDFLAGS=-L%PROJ_DIR%/%GPG_DLL%
    mingw32-make.exe bin/skopeo
    copy bin\skopeo %DIST_DIR%\skopeo.exe

artifacts:
  - path: dist
 ```
</details>

**Comment by gbraad** [comment]
> Are there any other aspects that rely on CGO?

@lsm5 ^^

Also, it might help to have community builds happening on appveyor to allow for testing Windows, but non-blocking/unreleased. WDYT?

**Comment by mtrmac** [comment]
> Does it make any sense to consider moving from the [mtrmac/gpgme](https://github.com/mtrmac/gpgme) gpgme wrapper that requires additional native libraries and CGO and instead moving to a native Go implementation?
> 
> From [golang/go#44226](https://github.com/golang/go/issues/44226):
> 
> > If you're looking for alternatives, consider the crypto/ed25519 package for simple signatures, golang.org/x/mod/sumdb/note for inline signatures, Tink for various cryptographic tools, or filippo.io/age for encryption (the latter two are not maintained by the Go team). You can read a summary of OpenPGP issues and alternatives here.
> 
> > If you are required to interoperate with OpenPGP systems and need a maintained package, we suggest considering a maintained community fork of golang.org/x/crypto/openpgp. We don't endorse any specific one.

Yeah, so using the maintained GpgME library to interact with a maintained `/usr/bin/gnupg` seems better than at least that one library above.

As for other Go implementations: It’s a code that could be integrated, but a significant subset (by no means _all_) users that strongly need signatures are also subject to various certification requirements like FIPS 140, which requires clearing of all memory containing private keys after it is no longer used, which is pretty much impossible to do purely in a garbage-collected language like Go, without specific language/compiler/runtime support for that, or without farming the key management to a different-language implementation.

I’d guess (without any data at all) that there are quite a lot of Windows users who want Skopeo for the ordinary `copy` functionality, not specifically for signing, so those could just use the verification-only non-go build. Is anyone here specifically interested in signing?

> Are there any other aspects that rely on CGO?

containers/storage device-mapper support (probably irrelevant for Windows), the `ostree` transport (probably irrelevant), PKCS#11 support for encrypted container images (probably relevant on Windows, but also somewhat niche).

**Comment by bofei222** [comment]
@xphyr have you made any further progress?  could  i  download a binary file for windows now ?

**Comment by gaborcsardi** [comment]
I needed a Windows binary on GitHub, so I used the excellent instructions at https://github.com/containers/skopeo/issues/715#issuecomment-917412078 to build one, and put it here: https://github.com/gaborcsardi/skopeo/releases

It is 64 bit with gpgme. It is just a zip-ped up skopeo directory that contains the required exe and dll files. The only difference to the instructions is that I did `strip -x` on the exe, to make it (much) smaller.

I did run a quick Windows 10 virus scan on it, but the usual disclaimers apply. It is probably better to run the build on your machine for yourself.

**Comment by mpils** [comment]
Thanks a lot @gaborcsardi for provided skopeo for windows release!
Here are some issued i was running into using skopeo in my cmd shell and related solution:

1) you have to use
`--override-os linux`        [mandatory]
`--override-arch amd64` [optional]
because skopeo use local os information overwise (os:windows) and doesn't find a container image supported for windows

`.\skopeo.exe --insecure-policy --override-os linux inspect docker://registry.fedoraproject.org/fedora`

for additional information
- you can find Golang $GOOS and $GOARCH supported values [here](https://go.dev/doc/install/source#environment)
- you can inspect container image supported values using skopeo (see example above)

2) you have to use
--insecure-policy     [mandatory]
because trust policy file /etc/containers/policy.json doesn't exist!

`.\skopeo --insecure-policy --override-os linux copy --additional-tag fedora:latest docker://registry.fedoraproject.org/fedora:latest docker-archive:fedora-latest.tar`

I hope somebody will find this usefull :)

**Comment by gaborcsardi** [comment]
@mpils Thanks!

1. I didn't notice this, because my use case needs `skopeo copy --all`.
2. An alternative is to use the `--policy` option to refer to a policy file explicitly. An example policy file:
    ```json
      {
         "default": [
           {
             "type": "insecureAcceptAnything"
           }
         ],
         "transports":
           {
             "docker-daemon":
               {
                 "": [{"type":"insecureAcceptAnything"}]
               }
           }
       }
    ```

**Comment by PeeZu** [comment]
Thanks @gaborcsardi for the Windows version, this save my life 👍 .

Any chance to have the latest Skopeo version Windows build on your fork ?

**Comment by gaborcsardi** [comment]
@PeeZu I just put v1.9.3 at https://github.com/gaborcsardi/skopeo/releases

However, this is even less tested than the previous version, so please do let me know if it works for you or not. Thanks!

**Comment by PeeZu** [comment]
Awesome @gaborcsardi ! 

First usage seems working as expected, thanks !

**Comment by gbraad** [comment]
@gaborcsardi So you manually performed these steps? As commented by Rickaro, it might be helpful to set up a Powershell script or a CI defintion: https://github.com/containers/skopeo/issues/715#issuecomment-943182133 and contribute this to the project. Even though they might not publish releases, it does show the build artifacts and build state.

**Comment by esemplare** [comment]
I've recently come across spokeo while needing to work with images from hosts without docker runtimes to pull. This is a great utility and part of this requirement is running the tool from a windows host.

I've successfully cross compiled from the main branch following @Rickaro 's instructions here: https://github.com/containers/skopeo/issues/715#issuecomment-917412078 with gpgme.

I'm only using it for the aspect of copy and sync, however I'm not sure what other tests to run to ensure there are issues with the other commands.

With the change to stop calling gpgme-config and using CGO overrides (https://github.com/containers/skopeo/commit/ebc5573e83fc97d2896a619af6a19d2981c8de74) the build failed due to a windows install of gpg4me not including a pkg-config file. I ended up creating one manually.

Once the project build, the other issue was that it would try to create invalid windows directories, in this case failing on a colon.

We could replace with something like "_" which is what curl does, or perhaps URL encoding (%3A) so the information of what was replaced is retained.

I ended up placing this in a runtime check for windows. I didn't spend too much time on coming up with an ideal solution but it worked for now.

```
case directory.Transport.Name():
    if runtime.GOOS == "windows" {
        destination = strings.Replace(destination, ":", "%3A", -1)
    }
```

---

All this said, I have some questions:

* @gbraad , do you know what issues I should expect regarding `io/fs` if any? I used Go `1.20.4 windows/amd64`.
* What are the thoughts on including platform specific code? Have a sync_windows.go ?
* Thoughts on the override-os comment @mpils mentioned? Should it default to linux regardless of the actual os?
* Locations for the files under `/etc/containers` ? Couldn't tell how those default locations are referenced.

**Comment by gbraad** [comment]
We did a small test with skopeo on Windows, but eventually decided to use different logic to deal with our requirement. An additional binary, built and maintained by us, was not the ideal path forward. So, unfortunately I can not say more than 'it worked', though our testing has been minimal. We might revisit this at some later time, though for now we have concluded and fulfilled our requirement differently.

Note: I work on CRC/OpenShift Local (and work on podman desktop/podman machine integrations) providing support on multiple platforms.

**Comment by mtrmac** [comment]
> * What are the thoughts on including platform specific code? Have a sync_windows.go ?

Yes, typically. As an alternative, there is _some_ precedent in c/image for using a run-time check based on comparing `runtime.GOOS`, which is more costly but has the benefit that the code can have unit tests that run on Linux as well… but that would only make a difference if that code _had_ unit tests (which is not the case for most of `skopeo sync` now).

> * Thoughts on the override-os comment @mpils mentioned? Should it default to linux regardless of the actual os?

Long-term, I’m not much of a fan — it would make it much harder to start using containers (or OCI artifacts) for non-Linux systems in the future. This equally, or perhaps even more, applies to macOS.

I appreciate that in a short-term horizon, defaulting the OS of images to copy to Linux is quite convenient. Maybe that can be achieved using a shell alias.

> * Locations for the files under `/etc/containers` ? Couldn't tell how those default locations are referenced.

https://github.com/containers/skopeo/issues/787 contains some pointers at the implementation mechanism.

Those `/etc` paths are clearly not the right defaults on Windows — _are there_ good defaults that can be used? I’m not sure Windows has an equivalent of a system-wide multi-user `/etc` like this, but I know very little about Windows. If there are good defaults, ideally we should add them to the containers/image codebase directly, so that they are used automatically.

**Comment by dhirschfeld** [comment]
> *I’m not sure Windows has an equivalent of a system-wide multi-user `/etc`*

That's probably `C:/ProgramData`

**Comment by passcod** [comment]
I've made a github repo that cross-compiles the non-gpg version, and puts it up in a release. That's sufficient for my needs, but would welcome a PR to build a better version. Does so once a week so it will always keep up to date with releases. https://github.com/passcod/winskopeo

**Comment by hurzelpurzel** [comment]
Thanks to @passcod  create a PR  Update install documentation #2124  to refer to your packackge repository

**Comment by Hollerweger** [comment]
> I've made a github repo that cross-compiles the non-gpg version, and puts it up in a release. That's sufficient for my needs, but would welcome a PR to build a better version. Does so once a week so it will always keep up to date with releases. https://github.com/passcod/winskopeo

Thanks! I tried the windows build but i get following issue with `skopeo login`:
`getaddrinfow: A non-recoverable error occurred during a database lookup`

Tested latest and also oldest version, same issue.

Maybe related to to a similar problem described here: https://stackoverflow.com/questions/32190073/getaddrinfo-a-non-recoverable-error-occurred-during-a-database-lookup

Same issue also with this version:

> [@PeeZu](https://github.com/PeeZu) I just put v1.9.3 at https://github.com/gaborcsardi/skopeo/releases
> 
> However, this is even less tested than the previous version, so please do let me know if it works for you or not. Thanks!


Running on Windows 11 x64

**Comment by Hollerweger** [comment]
Issue seems to be related to having the executable on a Network drive. See: https://github.com/exercism/cli/issues/168#issuecomment-204750203

Works now after copying it to a local windows folder

**Comment by Hollerweger** [comment]
Unfortunately `--override-os linux` or `--override-os=linux` is not working for me. It always pushes the image as windows container:
`skopeo copy --insecure-policy --override-os linux tarball:image.tar docker://repo:port/tag`

Got it working now with crane
`crane push --insecure --platform linux image.tar repo:port/tag`

**Comment by mtrmac** [comment]
@Hollerweger please file separate GitHub issues for separate problems; that has nothing to do with “Windows binary of Skopeo”.

**Comment by Hollerweger** [comment]
Of course it has to do with the windows binary as it sets the image to os windows and the override does not work for me.
See also remark here:

> Thanks a lot [@gaborcsardi](https://github.com/gaborcsardi) for provided skopeo for windows release! Here are some issues i was running into using skopeo in my cmd shell and related solution:
> 
> 1. you have to use
>    `--override-os linux`        [mandatory]
>    `--override-arch amd64` [optional]
>    because skopeo use local os information overwise (os:windows) and doesn't find a container image supported for windows

**Comment by gaborcsardi** [comment]
Well, in general if the os of the image is different from the local os, you'll have to use `--override-os`. E.g. if you are pushing a windows image from windows, then you don't need this. Same with the arch.

**Comment by Hollerweger** [comment]
> Well, in general if the os of the image is different from the local os, you'll have to use `--override-os`. E.g. if you are pushing a windows image from windows, then you don't need this. Same with the arch.

Yes clear, but override is not working as written. Tested with several windows releases provided in this chat.

**Comment by gaborcsardi** [comment]
> > Well, in general if the os of the image is different from the local os, you'll have to use `--override-os`. E.g. if you are pushing a windows image from windows, then you don't need this. Same with the arch.
> 
> Yes clear, but override is not working as written. Tested with several windows releases provided in this chat.

We can try to build a newer skopeo to see if this is fixed already.

**Comment by mtrmac** [comment]
This can’t be the issue to track “every single problem that happens when running Skopeo on Windows”. One issue per problem, please. Otherwise I’ll just close this eventually because a dozen-problem tracking issue is impossible to deal with.

“`tarball:` in the CLI is not working how I expect” equally applies to Linux, and is entirely distinct from “I don’t have a way to run Skopeo on Windows”.

**Comment by gaborcsardi** [comment]
PSA: I created a Windows build for v1.21.0, it is at https://github.com/gaborcsardi/skopeo/releases

Note that virustotal is not completely green, and I only tested it very lightly.

On the bright side, the process seems pretty easy now, so I'll try to write GHA workflow for it.

EDIT:  PR here: https://github.com/containers/skopeo/pull/2773
